### PR TITLE
BUG: arithmetic with reindex pow

### LINKED ISSUE #32685: Regression in artihmetic ops with alignment and special cases
I think 1.0.2 has a regression in arithmetic when we align but the result of `op(a, NA)` isn't NA. For example 1**np.nan should be 1.

Previous

```python
In [3]: pd.__version__
Out[3]: '1.0.1'

In [4]: pd.DataFrame({"A": [0, 1, 2]}) ** pd.DataFrame(index=[0, 1, 2])
Out[4]:
     A
0  NaN
1  1.0
2  NaN
```

1.0.2 / master

```python
In [1]: import pandas as pd

In [2]: pd.DataFrame({"A": [0, 1, 2]}) ** pd.DataFrame(index=[0, 1, 2])
Out[2]:
    A
0 NaN
1 NaN
2 NaN
```

Most likely from https://github.com/pandas-dev/pandas/pull/31679. cc @jbrockmendel if you have a chance to look (I won't today).

Are there any other special cases?

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #32685
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_arithmetic.py [hunk 1]
```diff
@@ -823,3 +823,13 @@ def test_align_frame(self):
...
         result = ts + half.take(np.random.permutation(len(half)))
         tm.assert_frame_equal(result, expected)
+
+
+def test_pow_with_realignment():
+    # GH#32685 pow has special semantics for operating with null values
+    left = pd.DataFrame({"A": [0, 1, 2]})
+    right = pd.DataFrame(index=[0, 1, 2])
+
+    result = left ** right
```
Comment (TomAugspurger) [inline pandas/tests/frame/test_arithmetic.py]: This tests `x ** nan  == x`.

Can you add a test for `nan ** 0 == 1`?

Perhaps parametrize over series & frame ops?

**Review by TomAugspurger — COMMENTED** [review]
Fix looks good. A couple comments on the test.

I have a PR open adding a 1.0.3.rst

**Comment by jreback** [comment]
lgtm. needs a rebase

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
@Meeseeksdev backport to 1.0.x

**Comment by TomAugspurger** [comment]
@jbrockmendel do you have time to backport this today?

**Comment by jbrockmendel** [comment]
> do you have time to backport this today?

yes

**Pull request merged**